src/uu/cp/src/platform/linux.rs

```diff
@@ -132,6 +135,48 @@ where
     Ok(num_bytes_copied)
 }
 
+#[cfg(any(target_os = "linux", target_os = "android"))]
+fn check_dest_is_fifo(dest: &Path) -> bool {
+    // If our destination file exists and it is a FIFO, we do a standard copy .
+    let file_type = fs::metadata(dest);
+    match file_type {
+        Ok(f) => f.file_type().is_fifo(),
+
+        _ => false,
+    }
+}
+
+#[cfg(any(target_os = "linux", target_os = "android"))]
+fn check_sparse_detection(source: &Path) -> Result<bool, std::io::Error> {
+    let mut src_file = File::open(source)?;
+
+    let metadata = src_file.metadata()?;
+    let size = metadata.size();
+    let blocks = metadata.blocks();
+    let blksize = metadata.blksize();
+    // cp uses a crude heuristic to detect sparse_files, an estimated formula which seems to accurately
+    // replicate that, is used. Modifications might be required if we observe any edges cases but this
+    // should work for majority of the files.
+    // Reference:https://doc.rust-lang.org/std/os/unix/fs/trait.MetadataExt.html#tymethod.blocks
+
+    let mut buf = vec![0; blksize as usize];
+
+    // Reading the file to account for virtual files which have 0 blocks allocated in disk
+    // and it is unknown if it contains data or not, virtual files need to be standard copied and
+    // categorized as not sparse.
+
+    if blocks == 0 {
+        let _ = src_file.read(&mut buf)?;
+
+        if buf.iter().any(|&x| x != 0x0) {
+            return Ok(false);
```

Comment:
1. If we have to treat virtual files specially, then it's probably a good idea to test them. Is it perhaps possible to write a test for this? Here's a wild idea, feel free to do something even better: `#[cfg(linux)]` and check that `du -B1 /sys/kernel/address_bits` outputs 0 and `cp --sparse-always /sys/kernel/address_bits foo` copies some content?
2. If we read some bytes and check them, then `return Ok(false);`, otherwise … still `return Ok(false);` (line 177). Doesn't this mean we can skip the reading completely?

Reply:
1.  I will look into creating a test , your idea seems really good . Instead of using ``du`` , we can just compare the metadata to get the blocks and check foo with lseek. If its preferable to use ``du`` , please tell me then.

2. Tbh , i am working locally on another PR , to fix the debug results . From there , some virtual files indeed would be detected as containing holes and debug results show they are sparse  , but contrary to the documented or intended behavior of cp (``--sparse=always`` ,`` --sparse=auto flags``), the copies are exactly as normal standard copies with a allocated desk space of the full size i.e. if we do ``cp --sparse=always /sys/kernel/profiling foo``  Foo will always have a block allocated. Thus i choose to categorize them as non-sparse here .

To explain , the condition ``blocks < size / 512`` is enough to detect sparse files . For any file that doesn't hold the condition , we can categorize as false . Any empty file created with ``truncate -sX a`` where X is any size >= MIN_BLOCK_SIZE (512) , also holds true to the condition in line 174 , so we get two types of zero blocks files with no disk allocation , ones with data and ones with no data . If we find data in it by reading,( we can reduce the buffer to MIN_BLOCK_SIZE to make it more performant ) we return Ok(false) , otherwise if there is no data in the zero block file, we return Ok(true) . Any file not caught in these conditions would be false .

 I can nest the ``block==0`` condition within ``blocks < size /512 `` if that makes it more readable .

Reply:
1. Doink! You're right, just fetching the metadata is a much more straight-forward approach, good idea.
2. Sorry, I completely misread the code. For some reason I overlooked the `else if blocks < size / 512` and thought it would only say `else {`. Nevermind, my bad! :D

So from my comments only "let's test virtual files" survives.

Reply:
I added a test , we don't really need to see what data has been copied . Any file with size greater than 0 contains data if we do a standard copy . If its preferable to read the file to make sure data is there , i can read the destination file instead but i don't believe its necessary since the copy won't behave like a virtual file.